File: contests/Week4/F.py
from collections import defaultdict

n = int(input())
graph = defaultdict(list)


def dfs(u, parent):
    val = 0
    for v in graph[u]:
        if parent == v:
            continue
        val += dfs(v, u)
    return val - 1 if val else val + 1


for _ in range(n-1):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

if dfs(1, 1):
    print("Alice")
else:
    print("Bob")

File: archive/contests/Week4/F.cpp
#include<iostream>
#include<vector>
#include<set>
#include<algorithm>
using namespace std;
#define endl "\n"

vector<int> graph[100005];

int dfs(int u = 1, int parent = -1) {
    int val = 0;
    for (int v : graph[u])
        if (parent != v)
            val += dfs(v, u);
    return val ? val - 1 : 1;
}

int main() {
    int n, u, v;
    cin >> n;
    for (int i = 0; i < n-1; i++)
        cin >> u >> v,
        graph[u].push_back(v),
        graph[v].push_back(u);
    cout << (dfs() ? "Alice" : "Bob") << endl;
    return 0;
}
